packages/cli/src/services/role.service.ts

```diff
@@ -0,0 +1,146 @@
+import { RoleRepository, SharedWorkflowRepository } from '@/databases/repositories';
+import { Service } from 'typedi';
+import { CacheService } from './cache.service';
+import type { Role, RoleNames, RoleScopes } from '@/databases/entities/Role';
+
+@Service()
+export class RoleService {
+	constructor(
+		private roleRepository: RoleRepository,
+		private sharedWorkflowRepository: SharedWorkflowRepository,
+		private cacheService: CacheService,
+	) {
+		void this.primeCache();
+	}
+
+	async primeCache() {
+		const allRoles = await this.roleRepository.find({});
+
+		if (!allRoles) return;
+
+		void this.cacheService.setMany<Role>(
+			allRoles.map((role) => [`cache:role:${role.scope}:${role.name}`, role]),
+		);
+	}
+
+	/**
+	 * role finders
+	 */
+
+	private async findCached(
+		scope: RoleScopes,
+		name: RoleNames,
+		options: { orFail: true },
+	): Promise<Role>;
+	private async findCached(
+		scope: RoleScopes,
+		name: RoleNames,
+		options: { orFail: false },
+	): Promise<Role | null>;
+	private async findCached(scope: RoleScopes, name: RoleNames, options: { orFail: boolean }) {
+		const cacheKey = `cache:role:${scope}:${name}`;
+
+		const cachedRole = await this.cacheService.get<Role>(cacheKey);
+
+		if (cachedRole) return this.roleRepository.create(cachedRole);
+
+		let dbRole: Role | null;
+
+		if (options.orFail) {
+			dbRole = await this.roleRepository.findRoleOrFail(scope, name);
+		} else {
+			dbRole = await this.roleRepository.findRole(scope, name);
+		}
+
+		if (dbRole !== null) {
+			void this.cacheService.set(cacheKey, dbRole);
+		}
+
+		return dbRole;
+	}
+
+	/**
+	 * global owner
+	 */
+
+	async findGlobalOwnerRole() {
+		return this.findCached('global', 'owner', { orFail: false });
+	}
+
+	async findGlobalOwnerRoleOrFail() {
```

Comment:
Love it, much simpler!

Reply:
Theoretically roles should never be missing - they're populated by migrations and for now should be just a hard coded list of 6 items.

We'll be implementing advanced permissions and we'll certainly revisit roles, so I don't think creating roles on the fly is a good idea as it's just a sign of malfunctioning code at this time.

Reply:
In a way they're still hardcoded - the typings for `name` and `scope` will prevent any new roles from being created. Should I revert be1ac57?

Reply:
The `name` and `scope` are not tightly coupled, right? Such as  `workflow + member` role, where `workflow` would be comprised only of `owner` and `editor`. If they're tightly coupled I don't think reverting is necessary, but if not, then I think removing the automatic creation is best to prevent us from messing up the code for now.

Reply:
Good point! Added validation to ensure we don't accidentally create a non-existing role: f581886. Also helps so we have a list of all the valid combinations.